base/src/main/java/org/aya/generic/Shaped.java

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2020-2022 Yinsen (Tesla) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.generic;
+
+import org.aya.core.def.CtorDef;
+import org.aya.core.repr.AyaShape;
+import org.aya.core.term.CallTerm;
+import org.aya.core.term.ErrorTerm;
+import org.aya.core.term.Term;
+import org.aya.generic.util.InternalException;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+public interface Shaped<T> {
+  @NotNull AyaShape shape();
+  @NotNull Term type();
+  @NotNull T constructorForm();
+  <O> boolean sameEncoding(@NotNull Shaped<O> other);
+
+  interface Inductively<T> extends Shaped<T> {
+    @Override @NotNull Term type();
+    @NotNull T makeZero(@NotNull CtorDef zero);
+    @NotNull T makeSuc(@NotNull CtorDef suc, @NotNull T t);
+    @NotNull T destruct(int repr);
```

Comment:
Did you mean `construct`?

Reply:
It means "when this function is called, current literal is destructed". So I think the name is fine.